pkg/storageops/gce/gce.go

```diff
@@ -0,0 +1,552 @@
+
+const googleDiskPrefix = "/dev/disk/by-id/google-"
+
+type gceOps struct {
+	inst    *Instance
+	service *compute.Service
+	mutex   sync.Mutex
+}
+
+// Instance stores the metadata of the running GCE instance
+type Instance struct {
+	ID         string
+	Name       string
+	Hostname   string
+	Zone       string
+	Project    string
+	InternalIP string
+	ExternalIP string
+	LBRequest  string
+	ClientIP   string
+	Error      string
+}
+
+type assigner struct {
+	err error
+}
+
+func (a *assigner) assign(getVal func() (string, error)) string {
+	if a.err != nil {
+		return ""
+	}
+
+	s, err := getVal()
+	if err != nil {
+		a.err = err
+	}
+
+	return s
+}
+
+func getEnvValueStrict(key string) (val string, err error) {
+	if val = os.Getenv(key); len(val) != 0 {
+		return
+	}
+
+	err = fmt.Errorf("env variable %s is not set", key)
+	return
+}
+
+// gceInfo fetches the GCE instance metadata from the metadata server
+func gceInfo(instance *Instance) {
+	a := &assigner{}
+	instance.ID = a.assign(metadata.InstanceID)
+	instance.Zone = a.assign(metadata.Zone)
+	instance.Name = a.assign(metadata.InstanceName)
+	instance.Hostname = a.assign(metadata.Hostname)
+	instance.Project = a.assign(metadata.ProjectID)
+	instance.InternalIP = a.assign(metadata.InternalIP)
+	instance.ExternalIP = a.assign(metadata.ExternalIP)
+
+	if a.err != nil {
+		instance.Error = a.err.Error()
+	}
+}
+
+func gceInfoFromEnv(instance *Instance) {
+	a := &assigner{}
+	instance.Name = a.assign(func() (string, error) { return getEnvValueStrict("GCE_INSTANCE_NAME") })
+	instance.Zone = a.assign(func() (string, error) { return getEnvValueStrict("GCE_INSTANCE_ZONE") })
+	instance.Project = a.assign(func() (string, error) { return getEnvValueStrict("GCE_INSTANCE_PROJECT") })
+
+	if a.err != nil {
+		instance.Error = a.err.Error()
+	}
+}
+
+// IsDevMode checks if the pkg is invoked in developer mode where GCE credentials
+// are set as env variables
+func IsDevMode() bool {
+	var i = new(Instance)
+	gceInfoFromEnv(i)
+	return len(i.Error) == 0
+}
+
+// NewClient creates a new GCE operations client
+func NewClient() (storageops.Ops, error) {
+	var i = new(Instance)
+	if metadata.OnGCE() {
+		gceInfo(i)
+	} else if ok := IsDevMode(); ok {
+		gceInfoFromEnv(i)
+	} else {
+		return nil, fmt.Errorf("instance is not running on GCE")
+	}
+
+	if len(i.Error) != 0 {
+		return nil, fmt.Errorf("error while fetching instance information. Err: %v", i.Error)
+	}
+
+	c, err := google.DefaultClient(context.Background(), compute.ComputeScope)
+	if err != nil {
+		return nil, fmt.Errorf("failed to authenticate with google api. Err: %v", err)
+	}
+
+	service, err := compute.New(c)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create Compute service: %v", err)
+	}
+
+	return &gceOps{
+		inst:    i,
+		service: service,
+	}, nil
+}
+
+func (s *gceOps) waitStatus(id string, desired string) error {
+	actual := ""
+	for retries, maxRetries := 0, 10; actual != desired && retries < maxRetries; retries++ {
```

Comment:
nit: constant for max retries. If possible same for both AWS and GCE.

Reply:
Also for all sleep intervals. Looks like they are common between AWS and GCE.

Reply:
Fixed in incremental